clients/src/main/java/org/apache/kafka/common/config/ConfigDef.java

```diff
@@ -136,27 +369,160 @@ public ConfigDef withClientSaslSupport() {
      * the appropriate type (int, string, etc)
      */
     public Map<String, Object> parse(Map<?, ?> props) {
+        return parse(props, true);
+    }
+
+    /**
+     * Parse and validate configs against this configuration definition. The input is a map of configs. It is expected
+     * that the keys of the map are strings, but the values can either be strings or they may already be of the
+     * appropriate type (int, string, etc). This will work equally well with either java.util.Properties instances or a
+     * programmatically constructed map.
+     *
+     * @param props The configs to parse and validate
+     * @param checkRequired Whether to throw a config exception in case that we don't provide a value for a required config
+     * @return Parsed and validated configs. The key will be the config name and the value will be the value parsed into
+     * the appropriate type (int, string, etc)
+     */
+    public Map<String, Object> parse(Map<?, ?> props, boolean checkRequired) {
+        ensureDependentsDefined();
         /* parse all known keys */
-        Map<String, Object> values = new HashMap<String, Object>();
+        Map<String, Object> values = new HashMap<>();
         for (ConfigKey key : configKeys.values()) {
-            Object value;
+            Object value = null;
             // props map contains setting - assign ConfigKey value
-            if (props.containsKey(key.name))
+            if (props.containsKey(key.name)) {
                 value = parseType(key.name, props.get(key.name), key.type);
```

Comment:
What happens if this throws an exception? Do we only get a partially parsed config? I was actually expecting `checkRequired` to protect against missing required configs and invalid configs. Otherwise won't we just never try to parse some fields?